cpp/include/cudf/detail/utilities/cuda_memcpy.hpp

```diff
@@ -17,41 +17,95 @@
 #pragma once
 
 #include <cudf/utilities/export.hpp>
+#include <cudf/utilities/span.hpp>
 
 #include <rmm/cuda_stream_view.hpp>
 
 namespace CUDF_EXPORT cudf {
 namespace detail {
 
+namespace impl {
+
 enum class host_memory_kind : uint8_t { PINNED, PAGEABLE };
 
+void cuda_memcpy_async(
+  void* dst, void const* src, size_t size, host_memory_kind kind, rmm::cuda_stream_view stream);
+
+}  // namespace impl
+
 /**
- * @brief Asynchronously copies data between the host and device.
+ * @brief Asynchronously copies data from host to device memory.
  *
  * Implementation may use different strategies depending on the size and type of host data.
  *
- * @param dst Destination memory address
- * @param src Source memory address
- * @param size Number of bytes to copy
- * @param kind Type of host memory
+ * @param dst Destination device memory
+ * @param src Source host memory
  * @param stream CUDA stream used for the copy
  */
-void cuda_memcpy_async(
-  void* dst, void const* src, size_t size, host_memory_kind kind, rmm::cuda_stream_view stream);
+template <typename T>
+void cuda_memcpy_async(device_span<T> dst, host_span<T const> src, rmm::cuda_stream_view stream)
+{
+  auto const is_pinned = src.is_device_accessible();
+  impl::cuda_memcpy_async(
+    dst.data(),
+    src.data(),
+    std::min(dst.size_bytes(), src.size_bytes()),
```

Comment:
> This might cause surprises to users if they unintentionally pass spans of different sizes. We could instead throw in this case.

Yes, this should be a runtime check and it should throw. If the caller wants to copy subspans, the caller can create subspans. Spans, as view types, are meant to make this easy.

Reply:
Sure.
Just one thing to keep in mind for this use case:
`host_span{my_hv.data(), subsize}` is not the same as `host_span{my_hv}.subspan(0, subsize)` because the first one will not know if it's pointing to pinned memory.

Reply:
done!